File: OCPickerView/OCPickerView/MPickerView/MPickerView.m
//
//  MPickerView.m
//  OCPickerView
//
//  Created by mal on 2017/12/14.
//  Copyright © 2017年 mal. All rights reserved.
//

#import "MPickerView.h"
#import "UIView+mframe.h"

static CGFloat const KCenterLineHeight = 0.5;
static CGFloat const KLineMargin = 1;

@interface MPickerView()<UIScrollViewDelegate>

@end

@implementation MPickerView

//MARK: - -----------------get & set
- (UIView *)lineView
{
    UIView *lineView = [[UIView alloc] initWithFrame:self.bounds];
    lineView.backgroundColor = [UIColor clearColor];
    lineView.m_height = self.rowHeight;
    lineView.m_centerY = self.m_height / 2;
    
    UIView *topline = [[UIView alloc] initWithFrame:CGRectMake(0, KLineMargin, self.m_width, KCenterLineHeight)];
    topline.userInteractionEnabled = NO;
    
    UIView *bottomline = [[UIView alloc] initWithFrame:CGRectMake(0, lineView.m_height - KLineMargin, self.m_width, KCenterLineHeight)];
    bottomline.userInteractionEnabled = NO;
    
    if ([self.m_delegate respondsToSelector:@selector(m_centerLineColor)])
    {
        topline.backgroundColor = [self.m_delegate m_centerLineColor];
        bottomline.backgroundColor = [self.m_delegate m_centerLineColor];
    }
    [lineView addSubview:topline];
    [lineView addSubview:bottomline];
    lineView.userInteractionEnabled = NO;
    return lineView;
}

- (UIScrollView *)scrollView
{
    UIScrollView *scrollView = [[UIScrollView alloc] init];
    scrollView.showsVerticalScrollIndicator = NO;
    scrollView.showsHorizontalScrollIndicator = NO;
    scrollView.m_height = self.m_height;
    scrollView.delegate = self;
    if ([self.m_delegate respondsToSelector:@selector(m_noOfSections)])
    {
        scrollView.m_width = self.m_width / (CGFloat)[self.m_delegate m_noOfSections];
    }
    UITapGestureRecognizer *tapGes = [[UITapGestureRecognizer alloc] initWithTarget:self action:@selector(tapScrollView:)];
    [scrollView addGestureRecognizer:tapGes];
    return scrollView;
}

- (NSMutableArray<NSMutableArray<UIView *> *> *)itemViewsArray
{
    if (_itemViewsArray == nil)
    {
        _itemViewsArray = [NSMutableArray array];
    }
    return _itemViewsArray;
}

- (NSMutableArray<UIScrollView *> *)scrollViewArray
{
    if (_scrollViewArray == nil)
    {
        _scrollViewArray = [NSMutableArray array];
    }
    return _scrollViewArray;
}

//MARK: - -----------------Event Handler
- (void)tapScrollView:(UITapGestureRecognizer *)tapGes
{
    UIView *gesView = tapGes.view;
    if ([gesView isKindOfClass:[UIScrollView class]])
    {
        UIScrollView *scrollView = (UIScrollView *)gesView;
        NSInteger section = [self.scrollViewArray indexOfObject:scrollView];
        CGPoint tapPoint = [tapGes locationInView:scrollView];
        if (section != NSNotFound)
        {
            NSMutableArray<UIView*> *itemsArray = self.itemViewsArray[section];
            [itemsArray enumerateObjectsUsingBlock:^(UIView * _Nonnull obj, NSUInteger idx, BOOL * _Nonnull stop) {
                
                if (CGRectContainsPoint(obj.frame, tapPoint))
                {
                    [self selectRow:idx isSection:section];
                    *stop = YES;
                }
            }];
        }
    }
}

//MARK: - -----------------init
+ (MPickerView *)mPickerViewWithFrame:(CGRect)frame delegate:(id<MPickerViewDelegate>)delegate
{
    MPickerView *pickerView = [[MPickerView alloc] initWithFrame:frame];
    pickerView.m_delegate = delegate;
    [pickerView setUpView];
    return pickerView;
}

- (void)setUpView
{
    self.rowHeight = self.m_delegate.m_rowHeight;
    NSInteger sectionNo = self.m_delegate.m_noOfSections;
    CGFloat sectionWidth = self.m_width / sectionNo;
    NSInteger addRows = (NSInteger)(self.m_height / self.rowHeight) - 1;
    for (NSInteger section = 0; section < sectionNo ; section++)
    {
        UIScrollView *scrollView = [self scrollView];
        scrollView.m_left = sectionWidth * section;
        NSInteger rows = [self.m_delegate m_noOfRowsInSection:section];
        CGFloat contentH = self.rowHeight *(rows + addRows);
        scrollView.contentSize = CGSizeMake(0, contentH);
        NSMutableArray *itemArray = [NSMutableArray array];
        for (int row = 0; row < rows; row++)
        {
            UIView *itemView = [self.m_delegate m_viewForRow:row inSection:section];
            itemView.m_width = sectionWidth;
            itemView.m_height = self.rowHeight;
            itemView.m_top = (row + addRows / 2) * self.rowHeight;
            [scrollView addSubview:itemView];
            [itemArray addObject:itemView];
        }
        [self.itemViewsArray addObject:itemArray];
        [self.scrollViewArray addObject:scrollView];
        [self addSubview:scrollView];
    }
    [self addSubview:[self lineView]];
}

//MARK: - UIScrollViewDelegate
- (void)scrollViewDidEndDecelerating:(UIScrollView *)scrollView
{
    [self fitContentOffSetWithScrollView:scrollView];
}

- (void)scrollViewDidEndDragging:(UIScrollView *)scrollView willDecelerate:(BOOL)decelerate
{
    if (decelerate == NO)
    {
        [self fitContentOffSetWithScrollView:scrollView];
    }
}

//MARK: - -----------------Private Method
- (void)fitContentOffSetWithScrollView:(UIScrollView *)scrollView
{
    CGFloat contentOffSetY = scrollView.contentOffset.y;
    NSInteger scrollToRow = (NSInteger)(contentOffSetY / self.rowHeight);
    NSInteger moreMove = (NSInteger)contentOffSetY % (NSInteger)self.rowHeight;
    if (moreMove > (NSInteger)(self.rowHeight / 2))
    {
        scrollToRow++;
    }
    NSInteger section = [self.scrollViewArray indexOfObject:scrollView];
    if (section != NSNotFound)
    {
        [self selectRow:scrollToRow isSection:section];
    }
}

//MARK: - -----------------Public Method
- (void)selectRow:(NSInteger)row isSection:(NSInteger)inSection
{
    if (self.scrollViewArray.count > inSection && self.itemViewsArray.count > inSection)
    {
        UIScrollView *scrollView = self.scrollViewArray[inSection];
        [scrollView setContentOffset:CGPointMake(0, row * self.rowHeight) animated:YES];
        if ([self.m_delegate respondsToSelector:@selector(m_didselectRow:section:itemArray:)])
        {
            [self.m_delegate m_didselectRow:row section:inSection itemArray:self.itemViewsArray[inSection]];
        }
    }
}

@end

File: SwiftPickerView/SwiftPickerView/MPickerView.swift
//
//  MPickerView.swift
//  SwiftStudy3.0
//
//  Created by mal on 16/11/24.
//  Copyright © 2016年 maliang. All rights reserved.
//

import UIKit

protocol MPickerViewProtrol {
    
    /// 返回pickerview的列数
    ///
    /// - Returns: 列数
    func m_noOfSections() -> Int
    
    /// 每一列的行数
    ///
    /// - Parameter section: 列
    /// - Returns: 行数
    func m_noOfRowsInSection(section:Int) -> Int
    
    /// 行高
    ///
    /// - Returns: 行高
    func m_rowHeight() -> CGFloat
    
    /// 类似于tableView的cellForRow,返回每一行显示的view
    ///
    /// - Parameters:
    ///   - row: view所在行
    ///   - inSection: view所在列
    /// - Returns: UIView
    func m_viewForRow(row:Int, inSection:Int) -> UIView
    
    /// 中间分割线颜色
    ///
    /// - Returns: UIColor
    func m_centerLineColor() -> UIColor
    
    /// 选中行回调
    ///
    /// - Parameters:
    ///   - row: 被选中的行
    ///   - section: 被选中的列
    ///   - itemArray: 被选中的[view]
    func m_selectRow(row:Int, section:Int, itemArray:[UIView])
}

class MPickerView: UIView,UIScrollViewDelegate {
    
    var mProtrol : MPickerViewProtrol?
    var rowHeight:CGFloat = 0.0
    lazy var scrollViewArray = [UIScrollView]()
    lazy var itemViewsArray  = [[UIView]]()
    
    //MARK: get方法
    var lineView:UIView{
        
        let lineView = UIView.init(frame: self.bounds)
        lineView.height = self.rowHeight
        lineView.centerY = self.height / 2
        let topline = UIView.init(frame: CGRect.init(x: 0, y: 0, width: self.width, height: 0.5))
        topline.isUserInteractionEnabled = false
        topline.backgroundColor = UIColor.black
        let bottomline = UIView.init(frame: CGRect.init(x: 0, y: lineView.height, width: self.width, height: 0.5))
        bottomline.isUserInteractionEnabled = false
        bottomline.backgroundColor = UIColor.black
        lineView.addSubview(topline)
        lineView.addSubview(bottomline)
        lineView.isUserInteractionEnabled = false
        return lineView
    }
    
    var scrollView:UIScrollView{
        
        let scrollView = UIScrollView()
        scrollView.showsVerticalScrollIndicator = false
        scrollView.showsHorizontalScrollIndicator = false
        scrollView.height = self.height
        let sectionNo = self.mProtrol!.m_noOfSections()
        scrollView.width = self.width / CGFloat(sectionNo)
        scrollView.delegate = self
        let tapGes = UITapGestureRecognizer.init(target: self, action:#selector(tapScrollView(ges:)))
        scrollView.addGestureRecognizer(tapGes)
        return scrollView
    }
    
    func tapScrollView(ges:UITapGestureRecognizer) {
        
        let scrollView = ges.view as! UIScrollView
        let section = self.scrollViewArray.index(of: scrollView)
        let tapPoint = ges.location(in: scrollView)

        let itemsArray = self.itemViewsArray[section!]
        for (idx,item) in itemsArray.enumerated(){
        
            if item.frame.contains(tapPoint){
                self.selectRow(row: idx, inSection: section!)
                break
            }
        }
    }
    
    //MARK: 初始化方法
    /// 初始化pickerview
    ///
    /// - Parameters:
    ///   - frame: height = 奇数行 * 行高
    ///   - m_protrol: 代理
    /// - Returns: MPickerView*
    class func mPickerViewWithFrame(frame:CGRect, m_protrol:MPickerViewProtrol) -> MPickerView{
    
        let mPicker = MPickerView.init(frame: frame)
        mPicker.mProtrol = m_protrol
        mPicker.setUpView()
        return mPicker
    }
    
    func setUpView() {
        
        self.rowHeight = self.mProtrol!.m_rowHeight()
        let sectionNo = self.mProtrol!.m_noOfSections()
        let sectionWidth = self.width / CGFloat(sectionNo)
        let addRows =  Int(self.height / self.rowHeight) - 1
        for section in 0...sectionNo - 1{
            
            let scrollView = self.scrollView
            scrollView.left = CGFloat(section) * sectionWidth
            let rows = self.mProtrol!.m_noOfRowsInSection(section: section)
            let contentH =  self.rowHeight * CGFloat(rows + addRows)
            scrollView.contentSize = CGSize.init(width: 0, height: contentH)
            var itemArray = [UIView]()
            for row in 0...rows - 1{
            
                let itemView = self.mProtrol!.m_viewForRow(row: row, inSection: section)
                itemView.width = sectionWidth
                itemView.height = self.rowHeight
                itemView.top = CGFloat(row + addRows / 2) * self.rowHeight
                scrollView.addSubview(itemView)
                itemArray.append(itemView)
            }
            self.itemViewsArray.append(itemArray)
            self.scrollViewArray.append(scrollView)
            self.addSubview(scrollView)
        }
        self.addSubview(self.lineView)
    }
    
    //MARK: scrollview代理
    func scrollViewDidEndDecelerating(_ scrollView: UIScrollView) {
        
        self.fitContentOffSet(scrollView: scrollView)
    }
    
    func scrollViewDidEndDragging(_ scrollView: UIScrollView, willDecelerate decelerate: Bool) {
        
        if decelerate == false {
            
            self.fitContentOffSet(scrollView: scrollView)
        }
    }
    
    //MARK: function
    func fitContentOffSet(scrollView:UIScrollView){
        
        let contentOffSetY = scrollView.contentOffset.y
        var scrollToRow = Int(contentOffSetY) / Int(self.rowHeight)
        let moreMove = Int(contentOffSetY) % Int(self.rowHeight)
        if moreMove > Int(self.rowHeight / 2){
            
            scrollToRow += 1
        }
        if let section = self.scrollViewArray.index(of: scrollView){
        
            self.selectRow(row: scrollToRow, inSection: section)
        }
    }
    
    func selectRow(row : Int, inSection : Int) {
        
        let scrollView = self.scrollViewArray[inSection]
        scrollView.setContentOffset(CGPoint.init(x: 0, y: CGFloat(row) * self.rowHeight), animated: true)
        self.mProtrol!.m_selectRow(row: row, section: inSection, itemArray: self.itemViewsArray[inSection])
    }
}
